src/v2/providers/schedule.ts

```diff
@@ -0,0 +1,198 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2022 Firebase
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import * as express from 'express';
+import { logger } from '../..';
+import { copyIfPresent } from '../../common/encoding';
+import { timezone } from '../../common/timezone';
+import { ManifestEndpoint, ManifestRequiredAPI } from '../../runtime/manifest';
+import * as options from '../options';
+import { HttpsFunction } from './https';
+
+/** @hidden */
+interface ScheduleArgs {
+  schedule: string;
+  timeZone?: timezone;
+  retryCount?: number;
+  maxRetryDuration?: string;
+  minBackoffDuration?: string;
+  maxBackoffDuration?: string;
+  maxDoublings?: number;
+  opts: options.GlobalOptions;
+}
+
+/** @internal */
+export function getOpts(args: string | ScheduleOptions): ScheduleArgs {
+  if (typeof args === 'string') {
+    return {
+      schedule: args,
+      opts: {},
+    };
+  }
+  const scheduleArgs: Partial<ScheduleArgs> = {
+    schedule: args.schedule,
+    timeZone: args.timeZone,
+    retryCount: args.retryCount,
+    maxRetryDuration: args.maxRetryDuration,
+    minBackoffDuration: args.minBackoffDuration,
+    maxBackoffDuration: args.maxBackoffDuration,
+    maxDoublings: args.maxDoublings,
+  };
+  const opts = { ...args };
+  delete (opts as any).schedule;
+  delete (opts as any).timeZone;
+  delete (opts as any).retryCount;
+  delete (opts as any).maxRetryDuration;
+  delete (opts as any).minBackoffDuration;
+  delete (opts as any).maxBackoffDuration;
+  delete (opts as any).maxDoublings;
+
+  scheduleArgs.opts = opts;
+  return scheduleArgs as ScheduleArgs;
```

Comment:
I'm not 100% sure if we need this separation logic. Can't we instead add `ScheduleOptions` to list of type accepted by `optionsToEndpoint`? Does that lead to some other error I'm not seeing here?

https://github.com/firebase/firebase-functions/blob/9b6da2d8d12c70914e5ca6258c50b56239e4ad88/src/v2/options.ts#L291

Reply:
I think it's also good to reduce code duplication! One thing that stuck out to me here is the number of times we have to repeat scheduled function configs.

One trick that I'm seeing in other namespace is that we just cast `ScheduledOptions` as `GlobalOptions` and pass that to `options.optionsToEndpoint`. Would that work here too? Did you think that was gross?

Reply:
updated the PR to just cast instead of creating a new object & deleting fields